### PR TITLE
Provide a custom error message from stripe

### LINKED ISSUE #277: More detailed Stripe errors [$50]
We actually get more  helpful feedback than we're showing. We should provide distinct error message for things like a card declining vs a server error

---
There is a bounty on this issue in [Bountysource](https://www.bountysource.com/issues/45293475-more-detailed-stripe-errorrs)

### LINKED ISSUE #283: AppCenter Closes When Trying to Pay
The AppCenter has closed when trying to pay for both Notes-Up and Spice-Up.

I had the issue originally with Notes-Up since that was the app I tried paying for first and eventually I was able to pay. There was no issue with the download of Notes-Up. I have not been able to pay for Spice-Up so I just clicked pay later and it installed without issue.

### PR DESCRIPTION
**Pull request by tintou**
Fixes #277 

Should also fix #283

### CODE REVIEW

**Comment by lewisgoddard** [comment]
I'd question these strings. "payment info number"? Why not just say card number?

**Comment by tintou** [comment]
Because there was never a mention of "card" so I assumed it was on-purpose ¯\_(ツ)_/¯

File: src/Dialogs/StripeDialog.vala [hunk 1]
```diff
@@ -427,5 +430,42 @@ public class AppCenter.Widgets.StripeDialog : Gtk.Dialog {
...
+        unowned string error_type = error_object.get_string_member ("type");
+        debug ("Stripe error type: %s", error_type);
+        switch (error_type) {
+            case "card_error":
+                unowned string error_code = error_object.get_string_member ("code");
+                debug ("Stripe error code: %s", error_code);
+                switch (error_code) {
+                    case "incorrect_number":
+                    case "invalid_number":
+                        return _("Please review your payment info number and try again.");
```
Comment (danrabbit) [inline src/Dialogs/StripeDialog.vala]: "The card number is incorrect. Please try again using the correct card number."

File: src/Dialogs/StripeDialog.vala [hunk 2]
```diff
@@ -427,5 +430,42 @@ public class AppCenter.Widgets.StripeDialog : Gtk.Dialog {
...
+                    case "invalid_expiry_month":
+                    case "invalid_expiry_year":
+                        return _("Please review your payment info expiry and try again.");
```
Comment (danrabbit) [inline src/Dialogs/StripeDialog.vala]: "The expiration year is invalid. Please try again using the correct expiration date."

File: src/Dialogs/StripeDialog.vala [hunk 3]
```diff
@@ -427,5 +430,42 @@ public class AppCenter.Widgets.StripeDialog : Gtk.Dialog {
...
+        debug ("Stripe error type: %s", error_type);
+        switch (error_type) {
+            case "card_error":
+                unowned string error_code = error_object.get_string_member ("code");
+                debug ("Stripe error code: %s", error_code);
+                switch (error_code) {
+                    case "incorrect_number":
+                    case "invalid_number":
+                        return _("Please review your payment info number and try again.");
+                    case "invalid_expiry_month":
```
Comment (danrabbit) [inline src/Dialogs/StripeDialog.vala]: "The expiration month is invalid. Please try again using the correct expiration date."

File: src/Dialogs/StripeDialog.vala [hunk 4]
```diff
@@ -427,5 +430,42 @@ public class AppCenter.Widgets.StripeDialog : Gtk.Dialog {
...
+                    case "invalid_expiry_year":
+                        return _("Please review your payment info expiry and try again.");
+                    case "incorrect_cvc":
+                    case "invalid_cvc":
+                        return _("Please review your payment info CVC and try again.");
```
Comment (danrabbit) [inline src/Dialogs/StripeDialog.vala]: "The CVC number is incorrect. Please try again using the correct CVC."

File: src/Dialogs/StripeDialog.vala [hunk 5]
```diff
@@ -427,5 +430,42 @@ public class AppCenter.Widgets.StripeDialog : Gtk.Dialog {
...
+                    case "expired_card":
+                        return _("The provided payment info has expired.");
```
Comment (danrabbit) [inline src/Dialogs/StripeDialog.vala]: "The card has expired. Please try again with a different card."

File: src/Dialogs/StripeDialog.vala [hunk 6]
```diff
@@ -427,5 +430,42 @@ public class AppCenter.Widgets.StripeDialog : Gtk.Dialog {
...
+                    case "processing_error":
+                        return _("An payment processing error occurred, please retry later. We apologize for any inconvenience caused.");
```
Comment (danrabbit) [inline src/Dialogs/StripeDialog.vala]: "An error occurred while processing the card. Please try again later. We apologize for any inconvenience caused."

**Comment by danrabbit** [comment]
According to @btkostner some of these errors might not ever be reported to AppCenter for security purposes. We should find out which ones and make sure we're in line with what we expect Houston to send to us

**Comment by tintou** [comment]
AppCenter is asking Stripe directly here, so we are sure to get the Stripe errors here. The Errors from Houston are considered as "Processing system error"

**Comment by tintou** [comment]
I also fixed a possible crash here because there were Gtk calls in a different thread which is forbidden by Gtk

File: src/Dialogs/StripeDialog.vala [hunk 7]
```diff
@@ -427,5 +437,43 @@ public class AppCenter.Widgets.StripeDialog : Gtk.Dialog {
 
         return data.str;
     }
+
+    private string get_stripe_error (Json.Object error_object) {
```
Comment (donadigo) [inline src/Dialogs/StripeDialog.vala]: You don't check if there's `type` member in the `error_object` so, this could lead to some fatals.

File: src/Dialogs/StripeDialog.vala [hunk 8]
```diff
@@ -427,5 +437,43 @@ public class AppCenter.Widgets.StripeDialog : Gtk.Dialog {
...
+        unowned string error_type = error_object.get_string_member ("type");
+        debug ("Stripe error type: %s", error_type);
+        switch (error_type) {
+            case "card_error":
+                unowned string error_code = error_object.get_string_member ("code");
```
Comment (donadigo) [inline src/Dialogs/StripeDialog.vala]: The same as with the type member. Check if `error_object` actually has the `code` member.

File: src/Dialogs/StripeDialog.vala (hunk 7 above)
Comment (donadigo) [inline src/Dialogs/StripeDialog.vala]: Can we make this static? Doesn't look like it's accessing any members from the instance. Also, this function should return an `unowned string` since `_` function you're using does return an `unowned`: https://valadoc.org/glib-2.0/GLib._.html

File: src/Dialogs/StripeDialog.vala [hunk 9]
```diff
@@ -360,41 +367,44 @@ public class AppCenter.Widgets.StripeDialog : Gtk.Dialog {
...
-
-                    if (root_object.has_member ("errors")) {
-                        error = true;
+                if (root_object != null && root_object.has_member ("id")) {
+                    var token_id = root_object.get_string_member ("id");
+                    string? houston_data = post_to_houston (stripe_key, app_id, token_id, email_entry.text, (amount * 100).to_string ());
+                    if (houston_data != null) {
+                        debug ("Houston data:%s", houston_data);
+                        parser.load_from_data (houston_data);
+                        root_object = parser.get_root ().get_object ();
```
Comment (donadigo) [inline src/Dialogs/StripeDialog.vala]: Both `get_root` and `get_object` could return `null`. Since this is already in trunk, I won't force you to check for that, but it would be nice to see this checked.

File: src/Dialogs/StripeDialog.vala [hunk 10]
```diff
@@ -360,41 +367,44 @@ public class AppCenter.Widgets.StripeDialog : Gtk.Dialog {
 
             var data = get_stripe_data (stripe_key, email_entry.text, (amount * 100).to_string (), card_number_entry.text, card_expiration_entry.text[0:2], year, card_cvc_entry.text);
             debug ("Stripe data:%s", data);
-            var error = false;
+            string error = null;
```
Comment (donadigo) [inline src/Dialogs/StripeDialog.vala]: This should be `string?` not `string`, the `error` is nullable.

**Comment by btkostner** [comment]
@tintou check the houston channel. We were debugging a crash related to threading. Is it the same issue?

File: src/Dialogs/StripeDialog.vala (hunk 9 above)
Comment (tintou) [inline src/Dialogs/StripeDialog.vala]: This part of the code needs some cleanup, but I would prefer it being done in another PR

**Comment by danrabbit** [comment]
Everything works as expected for me! Just waiting for Adam's review :)
